repository/src/main/java/org/alfresco/repo/event2/EventGenerator.java

```diff
@@ -571,115 +548,144 @@ protected boolean isTransactionCommitted()
+            for (Map.Entry<AssociationRef, PeerAssociationEventConsolidator> entry : consolidators.getPeerAssocs().entrySet())
             {
-                if (LOGGER.isTraceEnabled())
-                {
-                    LOGGER.trace("Ignoring node updated event as no fields have been updated: " + nodeRef);
-                }
-                return null;
+                sendEvent(entry.getKey(), entry.getValue());
             }
+        }
 
-            logEvent(event, consolidator.getEventTypes());
-            return event;
+        protected void sendEvent(NodeRef nodeRef, NodeEventConsolidator consolidator)
+        {
+            sendEvent(nodeRef, consolidator, nodeToEventEligibilityVerifier());
         }
 
         protected void sendEvent(ChildAssociationRef childAssociationRef, ChildAssociationEventConsolidator consolidator)
         {
-            EventInfo eventInfo = getEventInfo(AuthenticationUtil.getFullyAuthenticatedUser());
-            eventGeneratorQueue.accept(()-> createEvent(eventInfo, childAssociationRef, consolidator));
+            sendEvent(childAssociationRef, consolidator, childAssociationToEventEligibilityVerifier());
+        }
+
+        protected void sendEvent(AssociationRef peerAssociationRef, PeerAssociationEventConsolidator consolidator)
+        {
+            sendEvent(peerAssociationRef, consolidator, null);
+        }
+
+        /**
+         * Handles all kinds of events and sends them within dedicated transaction.
+         *
+         * @param entityReference reference to an entity (e.g. node, child association, peer association)
+         * @param eventConsolidator object encapsulating events occurred in a transaction
+         * @param entityToEventEligibilityVerifier allows to verify if entity is eligible to generate an even. If null no verification is necessary
+         * @param <REF> entity reference type
+         * @param <CON> event consolidator type - extension of {@link EventConsolidator}
+         */
+        private  <REF extends EntityRef, CON extends EventConsolidator<REF, ? extends Resource>> void sendEvent(
+            final REF entityReference, final CON eventConsolidator, final TriPredicate<REF, CON, EventInfo> entityToEventEligibilityVerifier)
+        {
+            final EventInfo eventInfo = getEventInfo(AuthenticationUtil.getFullyAuthenticatedUser());
+            transactionService.getRetryingTransactionHelper().doInTransaction((RetryingTransactionCallback<Void>) () -> {
+                eventSender.accept(() -> createEvent(entityReference, eventConsolidator, eventInfo, entityToEventEligibilityVerifier));
+                return null;
+            }, true, true);
         }
 
-        private RepoEvent<?> createEvent(EventInfo eventInfo, ChildAssociationRef childAssociationRef, ChildAssociationEventConsolidator consolidator)
+        private <REF extends EntityRef, CON extends EventConsolidator<REF, ? extends Resource>> Optional<RepoEvent<?>> createEvent(
```

Comment:
This is hard to read and debugging such a method is even worse. There is no JavaDoc either to explain all these generics params.

Reply:
JavaDoc describing mentioned generic params is present for above `sendEvent` method, but you are right that here it might be useful as well - JavaDoc added. Regarding the method readability it might be slightly worse due to generic parameters, but overall method doesn't differs much from the three previous implementations. Debugging experience should be the same IMHO.